uniffi_macros/src/custom.rs

```diff
@@ -2,96 +2,263 @@
-// implementation that the library supplies
-pub(crate) fn expand_ffi_converter_custom_type(
-    ident: &Ident,
-    builtin: &Path,
-    udl_mode: bool,
-) -> syn::Result<TokenStream> {
-    let impl_spec = tagged_impl_header("FfiConverter", ident, udl_mode);
-    let derive_ffi_traits = derive_all_ffi_traits(ident, udl_mode);
-    let name = ident_to_string(ident);
+use quote::{quote, ToTokens};
+use syn::{
+    braced,
+    parse::{Parse, ParseStream},
+    spanned::Spanned,
+    token::Brace,
+    Expr, ExprClosure, Pat, Token, Type,
+};
+
+pub struct CustomTypeArgs {
+    custom_type: Type,
+    uniffi_type: Type,
+    options: CustomTypeOptions,
+}
+
+impl Parse for CustomTypeArgs {
+    fn parse(input: ParseStream<'_>) -> syn::Result<Self> {
+        // Parse the custom / UniFFI type which are both required
+        let custom_type = input.parse()?;
+        input.parse::<Token![,]>()?;
+        let uniffi_type = input.parse()?;
+        let options = if input.peek(Token![,]) {
+            input.parse::<Token![,]>()?;
+
+            if input.peek(Brace) {
+                let content;
+                braced!(content in input);
+                content.parse()?
+            } else {
+                CustomTypeOptions::default()
+            }
+        } else {
+            CustomTypeOptions::default()
+        };
+        Ok(Self {
+            custom_type,
+            uniffi_type,
+            options,
+        })
+    }
+}
+
+/// Handle extra options for the custom type macro
+#[derive(Default)]
+pub struct CustomTypeOptions {
+    into_existing: Option<ConvertClosure>,
+    try_from_existing: Option<ConvertClosure>,
+    remote: Option<kw::remote>,
+}
+
+impl CustomTypeOptions {
+    fn is_remote(&self) -> bool {
+        // Until we land #2807, force all custom types to act as if `remote` was present.  This
+        // causes the generate code to only implement FfiConverter for the local tag, which is the
+        // current way of doing things.
+        true
+    }
+}
+
+impl Parse for CustomTypeOptions {
+    fn parse(input: ParseStream<'_>) -> syn::Result<Self> {
+        parse_comma_separated(input)
+    }
+}
+
+impl UniffiAttributeArgs for CustomTypeOptions {
+    fn parse_one(input: ParseStream<'_>) -> syn::Result<Self> {
+        let lookahead = input.lookahead1();
+        if lookahead.peek(kw::into_existing) {
+            input.parse::<kw::into_existing>()?;
+            input.parse::<Token![:]>()?;
+            Ok(Self {
+                into_existing: Some(input.parse()?),
+                ..Self::default()
+            })
+        } else if lookahead.peek(kw::try_from_existing) {
+            input.parse::<kw::try_from_existing>()?;
+            input.parse::<Token![:]>()?;
+            Ok(Self {
+                try_from_existing: Some(input.parse()?),
+                ..Self::default()
+            })
+        // Temporarily disabled until we land #2807 and get the new remote types system
```

Comment:
```suggestion
        // FIXME(#2807): Temporarily disabled until we land #2807 and get the new remote types system
```

makes it easy to grep for it.